File: src/primes.cpp
/** This code was at Google I/O 2012: 
Breaking the JavaScript Speed Limit with V8 
(http://www.youtube.com/watch?v=UJPdhx5zTaw) **/

#include <stdio.h>

class Primes {
	public:
		int getPrimeCount() const { return prime_count; }
		int getPrime(int i) const { return primes[i]; }
		void addPrime(int i) { primes[prime_count++] = i; }
		
		bool isPrimeDivisible(int candidate) {
			for (int i = 1; i < prime_count; ++i) {
				if ((candidate % primes[i]) == 0) return true;
			}
			return false;
		}
	private:
		volatile int prime_count;
		volatile int primes[50000];
};

int main() {
	Primes p;
	int c = 1;
	while (p.getPrimeCount() < 50000) {
		if (!p.isPrimeDivisible(c)) {
			p.addPrime(c);
		}
		c++;
	}
	printf("%d\n", p.getPrime(p.getPrimeCount() - 1));
}

File: src/primes.py


class Primes:
    def __init__(self):
        self.prime_count = 0
        self.primes = [0]*50000

    def getPrimeCount(self):
        return self.prime_count

    def getPrime(self, i):
        return self.primes[i]

    def addPrime(self, i):
        self.primes[self.prime_count] = i
        self.prime_count += 1

    def isPrimeDivisible(self, candidate):
        for i in range(1, self.prime_count - 1):
            if (candidate % self.primes[i]) == 0:
                return True
        return False

p = Primes()
c = 1
while p.getPrimeCount() < 50000:
    if not p.isPrimeDivisible(c):
        p.addPrime(c)
    c += 1

print(p.getPrime(p.getPrimeCount()-1))
